docs/proposal/20220608-capa-ami-github-action.md

```diff
@@ -0,0 +1,346 @@
+
+    ```json
+    {
+      "k8s_releases": {
+        "min1": "v1.24.1",
+        "min2": "v1.23.7",
+        "min3": "v1.22.10"
+      }
+    }
+    ```
+
+    The purpose of `AMIBuildConfig.json` is to store the Kubernetes release versions for which we need to build the AMIs, and once a pull request is created with updated Kubernetes release version(s) in this file for any of the keys, the Presubmit workflow is triggered which is responsible for building temporary CAPA AMIs for the purpose of testing.
+    \
+    The environment variables that can be set for this workflow along with their defaults are:
+    * **GITHUB_TOKEN**
+      \
+      Used for providing required permissions to the workflow environment.
+      \
+      Default value: `${{ steps.generate-token.outputs.token }}`
+    * **CAPA_ACTION_BASE_BRANCH**
+      \
+      The base branch for the pull request created by the periodic workflow.
+      \
+      Default value: `"main"`
+    * **CAPA_ACTION_HEAD_BRANCH**
+      \
+      The head branch for the pull request created by the periodic workflow.
+      \
+      Default value: `"capa-ami-action"`
+    * **AMI_BUILD_CONFIG_FILENAME**
+      \
+      Name of the file that stores latest Kubernetes release versions for building AMIs (located in directory `hack/tools/ami` within the repository).
+      \
+      Default value: `"AMIBuildConfig.json"`
+    * **CAPA_ACTION_PR_REVIEWERS**
+      \
+      Reviewers to be requested on the pull request created by the periodic workflow.
+      \
+      Default value: `"zeborg,zebhinav"`
+    * **CAPA_ACTION_PR_ASSIGNEES**
+      \
+      Users to be assigned to the pull request created by the periodic workflow.
+      \
+      Default value: `"zeborg,zebhinav"`
+
+
+  * `ami_presubmit.yaml` (Presubmit Workflow)
+    \
+    This workflow is triggered whenever a pull request is created to update `hack/tools/ami/AMIBuildConfig.json`. It fetches the Kubernetes release versions from `min1`, `min2` and `min3` in the file and checks whether AMIs for these Kubernetes versions already exist in the AWS account. If an AMI for one of these Kubernetes versions already exists, the `image-builder` process is skipped for that version, otherwise the image-builder builds the AMIs for the operating systems defined in `AMI_BUILD_SUPPORTED_OS` environment variable and replicates them to all the regions mentioned in the `AMI_BUILD_REGIONS` environment variable.
+    \
+    The environment variables that can be set for this workflow along with their defaults are:
+    * **AMI_BUILD_CONFIG_FILENAME**
+      \
+      Name of the file that stores latest Kubernetes release versions for building AMIs (located in directory `hack/tools/ami` within the repository).
+      Default value: `"AMIBuildConfig.json"`
+    * **AMI_BUILD_CONFIG_DEFAULTS**
```

Comment:
Can these default values be customized as and when required? If yes, then how can we do that? As sometimes there are few changes in image-builder and k8s which are incompatible, and hence we have to override few values, as image-builder releases happens quite infrequently nowadays.

Reply:
I believe we can set any of these values here as key-value pairs which are mentioned in this config file:
https://github.com/kubernetes-sigs/image-builder/blob/b98772b7cd577cd3e746796c325a6e4d63e32a5b/images/capi/packer/ami/packer.json#L145-L204

The format to store OS-specific variables for our implementation would be as mentioned here as an example:
https://github.com/zeborg/cluster-api-provider-aws/blob/main/hack/tools/ami/AMIBuildConfigDefaults.json

Reply:
Oh this is helpful, thanks @zeborg 🙂